theon/src/main.rs

```diff
@@ -13,21 +13,175 @@
 #![cfg_attr(not(test), no_main)]
 #![cfg_attr(not(test), no_std)]
 
+//!
+//! Theon has two primary responsibilities:
+//!
+//! 1. Taking an inventory of and initializing the hardware so
+//!    that when a description of the machine can be passed when
+//!    control is transferred to the supervisor, and so that the
+//!    system is in a known state.  For example, all processors
+//!    will be parked in the supervisor, running in a
+//!    fully-formed address space, etc.
+//! 2. Loading and initializing all of the consitute binaries
+//!    that make up the system.  That is, the binary images of
+//!    all segments and tasks will be loaded, and segment
+//!    initializers will be invoked.
+//!
+
 extern crate alloc;
 
 mod allocator;
 mod x86_64;
 
+use core::ops::Range;
+
+use arch::{VPageAddr, HPA, MIB, PF4K, V4KA};
+
+type Result<T> = core::result::Result<T, &'static str>;
+
+#[derive(Clone, Copy, Debug)]
+enum BinaryType {
+    Segment,
+    Task,
+}
+
+type BinaryMeta = (&'static str, HPA, BinaryType);
+
+const fn load_addr(offset: usize) -> HPA {
+    let addr = (64 + offset * 32) * MIB;
+    HPA::new(addr as u64)
+}
+const BINARY_IMAGE_MEMORY_SIZE: usize = 16 * MIB;
+
+const BINARY_TABLE: &[BinaryMeta] = &[
+    ("devices", load_addr(0), BinaryType::Segment),
+    ("global", load_addr(1), BinaryType::Segment),
+    ("memory", load_addr(2), BinaryType::Segment),
+    ("monitor", load_addr(3), BinaryType::Segment),
+    ("scheduler", load_addr(4), BinaryType::Segment),
+    ("supervisor", load_addr(5), BinaryType::Segment),
+    ("system", load_addr(6), BinaryType::Segment),
+    ("trace", load_addr(7), BinaryType::Segment),
+    ("vcpu", load_addr(8), BinaryType::Task),
+    ("vm", load_addr(9), BinaryType::Task),
+];
+const BINARY_LOAD_REGION_START: HPA = load_addr(0);
+const BINARY_LOAD_REGION_END: HPA = load_addr(BINARY_TABLE.len());
+
+fn end_addr() -> usize {
+    extern "C" {
+        static end: [u8; 0];
+    }
+    unsafe { end.as_ptr() as usize }
+}
+
+/// Main entry point for the loader.
+///
+/// When we enter `main()`, the first 4GiB of the
+/// physical address space are mapped R/W at
+/// arch::HYPER_BASE_VADDR.  Note that the memory
+/// regions that make up both the binary archive
+/// as well as our load regions are mapped within
+/// this region, so we can address them via pointers.
 #[cfg_attr(not(test), start, no_mangle)]
 pub extern "C" fn main(mbinfo_phys: u64) -> ! {
-    x86_64::init::start(mbinfo_phys);
+    let multiboot = x86_64::init::start(mbinfo_phys);
+    let crate::x86_64::multiboot1::InitInfo { memory_regions, regions, modules } = multiboot.info();
+    core::mem::drop(memory_regions);
+    uart::panic_println!("end = {:016x}", end_addr());
+    uart::panic_println!("regions: {:#x?}", regions);
+    let bins = modules.iter().find(|m| m.name == Some("bin.a")).expect("found 'bin.a' in modules");
```

Comment:
This archive loading stuff feels like it should be in its own function.

Reply:
I gave it a go, but it actually got pretty messy pretty fast. Would you object to keeping this as-is for now, and then refactoring in a later cleanup?

Reply:
OK.  Maybe add a TODO so we don't forget.

Reply:
Done.